alf/utils/reservoir_sampler.py

```diff
@@ -0,0 +1,153 @@
+# Copyright (c) 2019 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+from alf.utils.data_buffer import DataBuffer
+from alf.utils.nest_utils import get_nest_batch_size
+from alf.utils.conditional_ops import run_if
+
+
+class ReservoirSampler(DataBuffer):
+    """
+    A reservoir sampler for keeping K samples without replacement in a single
+    pass of the time horizon (O(1) complexity at every time step). See wiki for
+    its definition details:
+    https://en.wikipedia.org/wiki/Reservoir_sampling
+
+    Suppose the maximal time step is T, usually a reservoir sampler will keep
+    every item with the same probability of K/T. Here we make some changes so that
+    each item t can be kept with a probability roughly in proportional to
+    t^s, where s is an integer speed defined by the user.
+
+    You can use this sampler as a replay buffer. One big difference between this
+    sampler and a uniform replay buffer is that samples overflown by a replay
+    buffer will never be sampled again (probability hard cutting-off).
+
+    Probability computation:
+    - The probability of the t-th sample being selected in the set is (s+1)*K/t
+    - The probability of the t-th sample not being replaced by the sample at time
+      t+m is (1 - (s+1)*K/(t+m) * 1/K) = (t+m-s-1) / (t+m)
+    - The over probability of the t-th sample still being kept in the set at time
+      T is
+
+      (s+1)*K/t * (t+1-s-1)/(t+1) * (t+2-s-1)/(t+2) * ... * (T-s-1)/T
+
+      which is proportional to (t-1)(t-2)..(t-s) / ((T-1)(T-2)...(T-s)) and is
+      roughly proportional to t^s, when t,T >> (s+1)*K
+    """
+
+    def __init__(self,
+                 data_spec: tf.TensorSpec,
+                 K,
+                 speed=0,
```

Comment:
Feel we should make speed mean the same thing as the speed for AdaptiveAverager and AdaptiveNormalizer

Reply:
changed